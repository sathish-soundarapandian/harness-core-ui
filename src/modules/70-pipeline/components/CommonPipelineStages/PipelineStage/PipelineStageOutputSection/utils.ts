/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { AllowedTypes, getMultiTypeFromValue, MultiTypeInputType } from '@harness/uicore'
import type { FormikErrors, FormikProps } from 'formik'
import { get, isEmpty, set } from 'lodash-es'
import type { UseStringsReturn } from 'framework/strings'
import type { PipelineStageOutputs } from 'services/pipeline-ng'

export const MAX_LENGTH = 64

export interface PipelineStageOutputSectionProps {
  children: React.ReactElement
}

export interface PipelineStageOutputData {
  outputs: PipelineStageOutputs[]
}

export interface OutputPanelInputSetProps {
  template: PipelineStageOutputData
  path: string
  allowableTypes: AllowedTypes
  readonly?: boolean
}

export interface ConectedOutputPanelInputSetProps extends OutputPanelInputSetProps {
  formik: FormikProps<PipelineStageOutputData>
}

interface ValidateOutputPanelInputSetProp {
  data: PipelineStageOutputData
  template?: PipelineStageOutputData
  getString?: UseStringsReturn['getString']
}

export const validateOutputPanelInputSet = ({
  data,
  template,
  getString
}: ValidateOutputPanelInputSetProp): FormikErrors<PipelineStageOutputData> => {
  const errors: FormikErrors<PipelineStageOutputData> = { outputs: [] }
  data?.outputs?.forEach((output: PipelineStageOutputs, index: number) => {
    const currentVariableTemplate = get(template, `outputs[${index}].value`, '')
    if (isEmpty(output.value) && getMultiTypeFromValue(currentVariableTemplate) === MultiTypeInputType.RUNTIME) {
      set(errors, `outputs[${index}].value`, getString?.('fieldRequired', { field: output.name }))
    }
  })
  return errors
}
