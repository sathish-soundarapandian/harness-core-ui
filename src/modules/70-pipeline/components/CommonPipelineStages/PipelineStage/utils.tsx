/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

/* eslint-disable react/function-component-definition */
import React from 'react'
import type { CellProps, Renderer } from 'react-table'
import {
  Text,
  FontVariation,
  Layout,
  Icon,
  Color,
  Popover,
  ButtonSize,
  ButtonVariation,
  Button,
  IconName
} from '@harness/uicore'
import { defaultTo, get } from 'lodash-es'
import { Classes, PopoverInteractionKind, Position } from '@blueprintjs/core'
import { useParams } from 'react-router-dom'
import { String, useStrings } from 'framework/strings'
import type { PMSPipelineSummaryResponse } from 'services/pipeline-ng'
import { useAppStore } from 'framework/AppStore/AppStoreContext'
import { StoreType } from '@common/constants/GitSyncTypes'
import routes from '@common/RouteDefinitions'
import type { PipelineType, ProjectPathProps } from '@common/interfaces/RouteInterfaces'
import css from './PipelineStageMinimalMode.module.scss'

interface CodeSourceWrapper {
  textName: string
  iconName: IconName
  size: number
}

export const PipelineNameIdTagCell: Renderer<CellProps<PMSPipelineSummaryResponse>> = ({ row }) => {
  const data = row.original
  return (
    <Layout.Horizontal spacing="xsmall">
      <Layout.Vertical>
        <Text font={{ variation: FontVariation.H6 }} lineClamp={1} color={Color.BLACK}>
          {data?.name}
        </Text>
        <Text font={{ variation: FontVariation.BODY }} lineClamp={1} color={Color.GREY_600}>
          <String stringID="idLabel" vars={{ id: data?.identifier }} />
        </Text>
      </Layout.Vertical>
      {Object.keys(get(data, 'tags', {})).length > 0 && (
        <>
          <Icon name="main-tags" size={15} />
          <Text color={Color.GREY_600}>{Object.keys(get(data, 'tags', {})).length}</Text>
        </>
      )}
    </Layout.Horizontal>
  )
}

export const CodeSourceCell: Renderer<CellProps<PMSPipelineSummaryResponse>> = ({ row }) => {
  const { gitDetails } = row.original
  const { getString } = useStrings()
  const data = row.original
  const { isGitSyncEnabled: isGitSyncEnabledForProject, gitSyncEnabledOnlyForFF } = useAppStore()
  const isGitSyncEnabled = isGitSyncEnabledForProject && !gitSyncEnabledOnlyForFF
  const isRemote = data.storeType === StoreType.REMOTE || isGitSyncEnabled
  const inlineWrapper: CodeSourceWrapper = {
    textName: getString('inline'),
    iconName: 'repository',
    size: 10
  }
  const remoteWrapper: CodeSourceWrapper = {
    textName: getString('repository'),
    iconName: 'remote-setup',
    size: 12
  }

  return (
    <div className={css.codeSourceColumnContainer}>
      <Popover
        disabled={!isRemote}
        position={Position.TOP}
        interactionKind={PopoverInteractionKind.HOVER}
        className={Classes.DARK}
        content={
          <Layout.Vertical spacing="small" padding="large" className={css.contentWrapper}>
            <Layout.Horizontal spacing="small" flex={{ alignItems: 'center', justifyContent: 'start' }}>
              <Icon name="github" size={14} color={Color.GREY_200} />
              <Text color={Color.WHITE} font={{ variation: FontVariation.SMALL }}>
                {get(gitDetails, 'repoName', get(gitDetails, 'repoIdentifier'))}
              </Text>
            </Layout.Horizontal>
            {gitDetails?.filePath && (
              <Layout.Horizontal spacing="small" flex={{ alignItems: 'center', justifyContent: 'start' }}>
                <Icon name="remotefile" size={14} color={Color.GREY_200} />
                <Text color={Color.WHITE} font={{ variation: FontVariation.SMALL }}>
                  {gitDetails.filePath}
                </Text>
              </Layout.Horizontal>
            )}
          </Layout.Vertical>
        }
      >
        <div className={css.codeSourceColumn}>
          <Icon
            name={isRemote ? remoteWrapper.iconName : inlineWrapper.iconName}
            size={isRemote ? remoteWrapper.size : inlineWrapper.size}
            color={Color.GREY_600}
          />
          <Text margin={{ left: 'xsmall' }} font={{ variation: FontVariation.TINY_SEMI }} color={Color.GREY_600}>
            {isRemote ? remoteWrapper.textName : inlineWrapper.textName}
          </Text>
        </div>
      </Popover>
    </div>
  )
}

export const ViewPipelineButtonCell: Renderer<CellProps<PMSPipelineSummaryResponse>> = ({ row, column }) => {
  const { getString } = useStrings()
  const { accountId, module } = useParams<PipelineType<ProjectPathProps>>()
  const data = row.original
  const orgIdentifier = (column as any)?.orgIdentifier
  const projectIdentifier = (column as any)?.projectIdentifier

  const openPipelineInNewTab = (): void => {
    const pipelineStudioPath = routes.toPipelineStudio({
      projectIdentifier,
      orgIdentifier,
      accountId: defaultTo(accountId, ''),
      pipelineIdentifier: get(data, 'identifier', ''),
      module,
      repoIdentifier: get(data, 'gitDetails.repoIdentifier'),
      connectorRef: get(data, 'connectorRef'),
      repoName: get(data, 'gitDetails.repoName'),
      branch: get(data, 'gitDetails.branch'),
      storeType: get(data, 'storeType') as StoreType
    })
    window.open(`${window.location.origin}${window.location.pathname}#${pipelineStudioPath}`, '_blank')
  }

  return (
    <Button
      variation={ButtonVariation.SECONDARY}
      size={ButtonSize.SMALL}
      text={getString('common.viewText')}
      icon="main-view"
      onClick={e => {
        e.preventDefault()
        e.stopPropagation()
        openPipelineInNewTab()
      }}
    />
  )
}
