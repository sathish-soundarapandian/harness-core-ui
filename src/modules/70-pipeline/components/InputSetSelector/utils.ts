/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { IconName, SelectOption } from '@harness/uicore'
import type { EntityGitDetails, InputSetSummaryResponse } from 'services/pipeline-ng'
import css from './InputSetSelector.module.scss'

type InputSetLocal = InputSetSummaryResponse & SelectOption
export interface InputSetValue extends InputSetLocal {
  type: InputSetSummaryResponse['inputSetType']
  gitDetails?: EntityGitDetails
}
export interface ChildPipelineStageProps {
  childOrgIdentifier: string
  childProjectIdentifier: string
  usePortal?: boolean
  inputSetReferences?: string[]
}

export const getIconByType = (type: InputSetSummaryResponse['inputSetType']): IconName => {
  return type === 'OVERLAY_INPUT_SET' ? 'step-group' : 'yaml-builder-input-sets'
}

export const onDragStart = (event: React.DragEvent<HTMLLIElement>, row: InputSetValue): void => {
  event.dataTransfer.setData('data', JSON.stringify(row))
  event.currentTarget.classList.add(css.dragging)
}

export const onDragEnd = (event: React.DragEvent<HTMLLIElement>): void => {
  event.currentTarget.classList.remove(css.dragging)
}

export const onDragLeave = (event: React.DragEvent<HTMLLIElement>): void => {
  event.currentTarget.classList.remove(css.dragOver)
}

export const onDragOver = (event: React.DragEvent<HTMLLIElement>): void => {
  if (event.preventDefault) {
    event.preventDefault()
  }
  event.currentTarget.classList.add(css.dragOver)
  event.dataTransfer.dropEffect = 'move'
}
