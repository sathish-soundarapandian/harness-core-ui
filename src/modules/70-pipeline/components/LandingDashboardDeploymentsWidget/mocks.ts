/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { Error, Failure, ExecutionResponseDeploymentsStatsOverview } from 'services/dashboard-service'

const deploymentStatsResponse = {
  deploymentsStatsSummary: {
    countAndChangeRate: { count: 7, countChangeAndCountChangeRateInfo: { countChange: 0, countChangeRate: 0.0 } },
    failureRateAndChangeRate: { rate: 28.571428571428573, rateChangeRate: 0.0 },
    deploymentRateAndChangeRate: { rate: 0.23333333333333334, rateChangeRate: 0.0 },
    deploymentStats: [
      {
        time: 1633651200000,
        countWithSuccessFailureDetails: {
          count: 1,
          successCount: 0,
          failureCount: 1
        }
      },
      {
        time: 1633910400000,
        countWithSuccessFailureDetails: {
          count: 2,
          successCount: 2,
          failureCount: 0
        }
      },
      {
        time: 1633996800000,
        countWithSuccessFailureDetails: {
          count: 1,
          successCount: 1,
          failureCount: 0
        }
      },
      {
        time: 1634083200000,
        countWithSuccessFailureDetails: {
          count: 3,
          successCount: 2,
          failureCount: 1
        }
      }
    ]
  },
  mostActiveServicesList: {
    activeServices: [
      {
        serviceInfo: { serviceName: 'svc1', serviceIdentifier: 'svc1' },
        projectInfo: { projectIdentifier: 'ishant_test', projectName: 'ishant test' },
        orgInfo: { orgIdentifier: 'default', orgName: 'Default' },
        accountInfo: { accountIdentifier: 'kmpySmUISimoRrJL6NL73w' },
        countWithSuccessFailureDetails: {
          count: 5,
          countChangeAndCountChangeRateInfo: { countChange: 0, countChangeRate: 0.0 },
          successCount: 4,
          failureCount: 1
        }
      }
    ]
  }
}

export const deploymentStatsSummaryResponse: {
  data: ExecutionResponseDeploymentsStatsOverview
  error: Error | Failure | undefined
  loading: boolean
} = {
  error: undefined,
  loading: false,
  data: {
    response: {
      deploymentsOverview: {
        runningExecutions: [],
        pendingApprovalExecutions: [],
        pendingManualInterventionExecutions: [],
        failed24HrsExecutions: []
      },
      ...deploymentStatsResponse
    },
    executionStatus: 'SUCCESS',
    executionMessage: 'Successfully fetched data'
  }
}

export const noDeploymentData: {
  data: ExecutionResponseDeploymentsStatsOverview
  error: Error | Failure | undefined
  loading: boolean
} = {
  error: undefined,
  loading: false,
  data: {
    response: {
      deploymentsOverview: {
        runningExecutions: [],
        pendingApprovalExecutions: [],
        pendingManualInterventionExecutions: [],
        failed24HrsExecutions: []
      },
      deploymentsStatsSummary: {
        countAndChangeRate: { count: 0, countChangeAndCountChangeRateInfo: { countChange: 0, countChangeRate: 0.0 } },
        failureRateAndChangeRate: { rate: 0, rateChangeRate: 0.0 },
        deploymentRateAndChangeRate: { rate: 0, rateChangeRate: 0.0 },
        deploymentStats: []
      },
      mostActiveServicesList: {
        activeServices: []
      }
    },
    executionStatus: 'SUCCESS',
    executionMessage: 'Successfully fetched data'
  }
}

export const deploymentStatsSummaryWithOverviewResponse = {
  status: 'SUCCESS',
  data: {
    response: {
      deploymentsOverview: {
        runningExecutions: [
          {
            pipelineInfo: {
              pipelineName: 'pipwithenvoverride',
              pipelineIdentifier: 'pipwithenvoverride'
            },
            projectInfo: {
              projectIdentifier: 'test1',
              projectName: 'test1'
            },
            orgInfo: {
              orgIdentifier: 'default',
              orgName: 'default'
            },
            accountInfo: {
              accountIdentifier: 'px7xd_BFRCi-pfWPYXVjvw'
            },
            planExecutionId: 'jVYFTPFkR4yv5IjOra9m-A',
            startTs: 1670849558369
          },
          {
            pipelineInfo: {
              pipelineName: 'Triggers - 2',
              pipelineIdentifier: 'Triggers_2'
            },
            projectInfo: {
              projectIdentifier: 'test2',
              projectName: 'test2'
            },
            orgInfo: {
              orgIdentifier: 'default',
              orgName: 'default'
            },
            accountInfo: {
              accountIdentifier: 'px7xd_BFRCi-pfWPYXVjvw'
            },
            planExecutionId: '6SMlruHPRl2Rzhuw7JP-LA',
            startTs: 1665166801126
          },
          {
            pipelineInfo: {
              pipelineName: 'Triggers - 2',
              pipelineIdentifier: 'Triggers_2'
            },
            projectInfo: {
              projectIdentifier: 'test2',
              projectName: 'test2'
            },
            orgInfo: {
              orgIdentifier: 'default',
              orgName: 'default'
            },
            accountInfo: {
              accountIdentifier: 'px7xd_BFRCi-pfWPYXVjvw'
            },
            planExecutionId: 'AjWas2yRReWoX1bncorSWg',
            startTs: 1665319801356
          }
        ],
        pendingApprovalExecutions: [],
        pendingManualInterventionExecutions: [],
        failed24HrsExecutions: [
          {
            pipelineInfo: {
              pipelineName: 'bug-trigger-update-after-pipeline-update',
              pipelineIdentifier: 'bugtriggerupdateafterpipelineupdate'
            },
            projectInfo: {
              projectIdentifier: 'testname',
              projectName: 'testname'
            },
            orgInfo: {
              orgIdentifier: 'default',
              orgName: 'default'
            },
            accountInfo: {
              accountIdentifier: 'px7xd_BFRCi-pfWPYXVjvw'
            },
            planExecutionId: 'xyTsxDgUSLaiK3baCcK9AQ',
            startTs: 1671292203048
          }
        ]
      },
      ...deploymentStatsResponse
    },
    executionStatus: 'SUCCESS',
    executionMessage: 'Successfully fetched data'
  },
  metaData: null,
  correlationId: null
}
