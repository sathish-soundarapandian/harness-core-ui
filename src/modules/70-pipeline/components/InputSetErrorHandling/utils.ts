/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import type { MutateMethod } from 'restful-react'
import { defaultTo, get } from 'lodash-es'
import { useHistory, useParams } from 'react-router-dom'
import { useToaster } from '@harness/uicore'
import type { InputSetDTO, SaveInputSetDTO, StatusType } from '@pipeline/utils/types'
import { AppStoreContext } from 'framework/AppStore/AppStoreContext'
import type {
  EntityGitDetails,
  ResponseInputSetResponse,
  ResponseOverlayInputSetResponse,
  UpdateInputSetForPipelinePathParams,
  UpdateInputSetForPipelineQueryParams,
  UpdateOverlayInputSetForPipelinePathParams,
  UpdateOverlayInputSetForPipelineQueryParams
} from 'services/pipeline-ng'
import type { SaveToGitFormInterface } from '@common/components/SaveToGitForm/SaveToGitForm'
import { StoreMetadata, StoreType } from '@common/constants/GitSyncTypes'
import { UseSaveSuccessResponse, useSaveToGitDialog } from '@common/modals/SaveToGitDialog/useSaveToGitDialog'
import type { InputSetGitQueryParams, InputSetPathProps, PipelineType } from '@common/interfaces/RouteInterfaces'
import { useQueryParams } from '@common/hooks'
import type { SaveToGitFormV2Interface } from '@common/components/SaveToGitFormV2/SaveToGitFormV2'
import { useStrings } from 'framework/strings'
import { yamlStringify } from '@common/utils/YamlHelperMethods'
import { clearNullUndefined } from '@pipeline/utils/inputSetUtils'
import useRBACError from '@rbac/utils/useRBACError/useRBACError'
import type {
  OverlayInputSetDTO,
  SaveOverlayInputSetDTO
} from '@pipeline/components/OverlayInputSetForm/OverlayInputSetForm'

type UpdateInpSetOrOverlayInpSetReturnType = Promise<{
  status?: StatusType
  nextCallback: () => void
}>

interface UseSaveInputSetOrOverlayInpSetReturnType {
  handleSubmit: (inputSetObj: InputSetDTO | OverlayInputSetDTO, storeMetadata?: StoreMetadata) => Promise<void>
}

interface InputSetInfo {
  updateInputSet: MutateMethod<
    ResponseInputSetResponse,
    string,
    UpdateInputSetForPipelineQueryParams,
    UpdateInputSetForPipelinePathParams
  >
  updateOverlayInputSet: MutateMethod<
    ResponseOverlayInputSetResponse,
    void,
    UpdateOverlayInputSetForPipelineQueryParams,
    UpdateOverlayInputSetForPipelinePathParams
  >
  inputSet: InputSetDTO | OverlayInputSetDTO
  _isOverlayInputSet: boolean
  refetch?: () => Promise<void>
  fromInputSetForm?: boolean
  hideForm?: () => void
  inpSetGitDetails?: EntityGitDetails
  onReconcile?: (identifier: string) => void
  refetchInputSets?: () => void
}

export function useSaveInputSetOrOverlayInpSet(inputSetInfo: InputSetInfo): UseSaveInputSetOrOverlayInpSetReturnType {
  const {
    updateInputSet,
    updateOverlayInputSet,
    inputSet,
    _isOverlayInputSet,
    refetch,
    fromInputSetForm,
    hideForm,
    inpSetGitDetails,
    onReconcile,
    refetchInputSets
  } = inputSetInfo
  const { getString } = useStrings()
  const { showSuccess, showError } = useToaster()
  const { getRBACErrorMessage } = useRBACError()
  const history = useHistory()
  const { isGitSyncEnabled } = React.useContext(AppStoreContext)

  const { projectIdentifier, orgIdentifier, accountId, pipelineIdentifier } = useParams<
    PipelineType<InputSetPathProps> & { accountId: string }
  >()
  const { repoIdentifier, branch, repoName, connectorRef, storeType } = useQueryParams<InputSetGitQueryParams>()

  const [savedInputSetObj, setSavedInputSetObj] = React.useState<InputSetDTO | OverlayInputSetDTO>({})
  const [initialGitDetails, setInitialGitDetails] = React.useState<EntityGitDetails>({ repoIdentifier, branch })
  const [initialStoreMetadata, setInitialStoreMetadata] = React.useState<StoreMetadata>({
    repoName,
    branch,
    connectorRef,
    storeType
  })

  const isInline = !isGitSyncEnabled && initialStoreMetadata.storeType !== StoreType.REMOTE

  /* istanbul ignore next */ const handleReconcile = (): void => {
    if (initialStoreMetadata.storeType === StoreType.REMOTE) {
      onReconcile?.(get(inputSet, 'identifier', ''))
    }
    refetch?.()
    hideForm?.()
    refetchInputSets?.()
  }

  const updateInpSetOrOverlayInpSet = React.useCallback(
    async (
      inputSetObj: InputSetDTO | OverlayInputSetDTO,
      gitDetails?: SaveToGitFormInterface,
      objectId = ''
    ): UpdateInpSetOrOverlayInpSetReturnType => {
      let response: ResponseInputSetResponse | ResponseOverlayInputSetResponse | null = null
      try {
        /* istanbul ignore else */ if (inputSetObj.identifier) {
          if (_isOverlayInputSet) {
            response = await updateOverlayInputSet(
              yamlStringify({ overlayInputSet: clearNullUndefined(inputSetObj) }) as unknown as void,
              {
                pathParams: {
                  inputSetIdentifier: defaultTo(inputSetObj.identifier, '')
                },
                queryParams: {
                  accountIdentifier: accountId,
                  orgIdentifier,
                  pipelineIdentifier: pipelineIdentifier ?? get(inputSet, 'pipelineIdentifier'),
                  projectIdentifier,
                  ...(isGitSyncEnabled
                    ? {
                        pipelineRepoID: repoIdentifier,
                        pipelineBranch: branch
                      }
                    : {}),
                  ...(initialStoreMetadata.storeType === StoreType.REMOTE ? initialStoreMetadata : {}),
                  ...(gitDetails
                    ? { ...gitDetails, lastObjectId: objectId, lastCommitId: initialGitDetails.commitId }
                    : {}),
                  ...(gitDetails && gitDetails.isNewBranch ? { baseBranch: initialGitDetails.branch } : {})
                }
              }
            )
          } else {
            response = await updateInputSet(yamlStringify({ inputSet: clearNullUndefined(inputSetObj) }), {
              pathParams: {
                inputSetIdentifier: defaultTo(inputSetObj.identifier, '')
              },
              queryParams: {
                accountIdentifier: accountId,
                orgIdentifier,
                pipelineIdentifier: pipelineIdentifier ?? get(inputSet, 'pipelineIdentifier'),
                projectIdentifier,
                ...(isGitSyncEnabled
                  ? {
                      pipelineRepoID: repoIdentifier,
                      pipelineBranch: branch
                    }
                  : {}),
                ...(initialStoreMetadata.storeType === StoreType.REMOTE ? initialStoreMetadata : {}),
                ...(gitDetails
                  ? { ...gitDetails, lastObjectId: objectId, lastCommitId: initialGitDetails.commitId }
                  : {}),
                ...(gitDetails && gitDetails.isNewBranch ? { baseBranch: initialGitDetails.branch } : {})
              }
            })
          }
          /* istanbul ignore else */ if (get(response, 'data') && isInline) refetch?.()
        } else {
          throw new Error(getString('common.validation.identifierIsRequired'))
        }
        showSuccess(
          _isOverlayInputSet
            ? getString('pipeline.inputSets.overlayISUpdated', { name: get(inputSet, 'name', '') })
            : getString('pipeline.inputSets.inputSetUpdated', { name: get(inputSet, 'name', '') })
        )
        /* istanbul ignore else */ if (isInline) {
          !!fromInputSetForm && history.goBack()
          hideForm?.()
          refetchInputSets?.()
        }
      } catch (e) {
        if (isInline) {
          showError(getRBACErrorMessage(e), undefined, 'pipeline.common.error')
        } else {
          throw e
        }
      }
      return {
        status: response?.status,
        nextCallback: /* istanbul ignore next */ () => (fromInputSetForm ? history.goBack() : handleReconcile())
      }
    },
    [
      _isOverlayInputSet,
      inputSet,
      isGitSyncEnabled,
      initialStoreMetadata,
      refetch,
      updateOverlayInputSet,
      accountId,
      orgIdentifier,
      pipelineIdentifier,
      projectIdentifier,
      repoIdentifier,
      branch,
      initialGitDetails.commitId,
      initialGitDetails.branch,
      updateInputSet,
      fromInputSetForm,
      history,
      hideForm,
      refetchInputSets
    ]
  )

  const { openSaveToGitDialog } = useSaveToGitDialog<SaveInputSetDTO | SaveOverlayInputSetDTO>({
    onSuccess: (
      gitData: SaveToGitFormInterface & SaveToGitFormV2Interface,
      payload?: SaveInputSetDTO | SaveOverlayInputSetDTO,
      objectId?: string
    ): Promise<UseSaveSuccessResponse> =>
      updateInpSetOrOverlayInpSet(
        defaultTo(
          _isOverlayInputSet
            ? (payload as SaveOverlayInputSetDTO)?.overlayInputSet
            : (payload as SaveInputSetDTO)?.inputSet,
          savedInputSetObj
        ),
        gitData,
        objectId
      )
  })

  const handleSubmit = React.useCallback(
    async (inputSetObj: InputSetDTO | OverlayInputSetDTO, storeMetadata?: StoreMetadata) => {
      const _gitDetails = fromInputSetForm ? get(inputSet, 'gitDetails') : inpSetGitDetails
      setSavedInputSetObj(inputSetObj)
      setInitialGitDetails(defaultTo(_gitDetails, {}))
      setInitialStoreMetadata(defaultTo(storeMetadata, {}))

      /* istanbul ignore else */ if (inputSetObj) {
        _isOverlayInputSet && delete inputSetObj.pipeline
        if (isGitSyncEnabled || get(storeMetadata, 'storeType') === StoreType.REMOTE) {
          openSaveToGitDialog({
            isEditing: true,
            resource: {
              type: 'InputSets',
              name: inputSetObj.name as string,
              identifier: inputSetObj.identifier as string,
              gitDetails: _gitDetails,
              storeMetadata: get(storeMetadata, 'storeType') === StoreType.REMOTE ? storeMetadata : undefined
            },
            payload: _isOverlayInputSet ? { overlayInputSet: inputSetObj } : { inputSet: inputSetObj }
          })
        } else {
          updateInpSetOrOverlayInpSet(inputSetObj)
        }
      }
    },
    [
      _isOverlayInputSet,
      isGitSyncEnabled,
      inputSet,
      updateInpSetOrOverlayInpSet,
      openSaveToGitDialog,
      inpSetGitDetails,
      fromInputSetForm
    ]
  )

  return {
    handleSubmit
  }
}
