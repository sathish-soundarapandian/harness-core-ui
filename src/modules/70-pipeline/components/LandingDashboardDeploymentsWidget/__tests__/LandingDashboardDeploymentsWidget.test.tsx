/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import routes from '@common/RouteDefinitions'
import * as dashboardService from 'services/dashboard-service'
import { deploymentStatsSummaryResponse, noDeploymentData } from '../mocks'
import LandingDashboardDeploymentsWidget from '../LandingDashboardDeploymentsWidget'

jest.mock('services/dashboard-service', () => ({
  useGetDeploymentStatsOverview: jest.fn().mockImplementation(() => {
    return { data: deploymentStatsSummaryResponse, refetch: jest.fn(), error: null, loading: false }
  })
}))

describe('LandingDashboardDeploymentsWidget tests', () => {
  test('cd dashboard charts should render properly', async () => {
    const { getByText } = render(
      <TestWrapper path={routes.toLandingDashboard({ accountId: ':accountId' })} pathParams={{ accountId: 'dummy' }}>
        <LandingDashboardDeploymentsWidget />
      </TestWrapper>
    )

    // expect(container).toMatchSnapshot()

    // const pendingApprovalsBadge = getByText('4 Pending Approvals')
    // expect(pendingApprovalsBadge).toBeInTheDocument()
    // const pendingManualInterventionsBadge = getByText('1 Pending Manual Interventions')
    // expect(pendingManualInterventionsBadge).toBeInTheDocument()

    const deploymentStatsSummaryCard = getByText('pipeline.dashboards.pipelineExecution')
    expect(deploymentStatsSummaryCard).toBeInTheDocument()
    const failureRateSummaryCard = getByText('common.failureRate')
    expect(failureRateSummaryCard).toBeInTheDocument()
    const deploymentFrequencySummaryCard = getByText('pipeline.executionFrequency')
    expect(deploymentFrequencySummaryCard).toBeInTheDocument()

    const mostActiveServicesTitle = getByText('common.mostActiveServices')
    expect(mostActiveServicesTitle).toBeInTheDocument()

    const successLegendText = getByText('Success (5)')
    expect(successLegendText).toBeInTheDocument()
    const failedLegendText = getByText('Failed (2)')
    expect(failedLegendText).toBeInTheDocument()
  })

  test('loading true', () => {
    jest.spyOn(dashboardService, 'useGetDeploymentStatsOverview').mockImplementation((): any => {
      return {
        data: [],
        refetch: jest.fn(),
        error: null,
        loading: true
      }
    })

    const { container } = render(
      <TestWrapper path={routes.toLandingDashboard({ accountId: ':accountId' })} pathParams={{ accountId: 'dummy' }}>
        <LandingDashboardDeploymentsWidget />
      </TestWrapper>
    )

    //spinner should be visible
    expect(container.querySelector('[data-icon="spinner"]')).toBeTruthy()
  })

  test('error true', () => {
    jest.spyOn(dashboardService, 'useGetDeploymentStatsOverview').mockImplementation((): any => {
      return {
        data: [],
        refetch: jest.fn(),
        error: true,
        loading: false
      }
    })

    const { getByText } = render(
      <TestWrapper path={routes.toLandingDashboard({ accountId: ':accountId' })} pathParams={{ accountId: 'dummy' }}>
        <LandingDashboardDeploymentsWidget />
      </TestWrapper>
    )

    // error should be visible
    expect(getByText('projectsOrgs.apiError')).toBeTruthy()
  })

  test('no deployments', () => {
    jest.spyOn(dashboardService, 'useGetDeploymentStatsOverview').mockImplementation((): any => {
      return {
        data: noDeploymentData,
        refetch: jest.fn(),
        error: null,
        loading: false
      }
    })

    const { getByText } = render(
      <TestWrapper path={routes.toLandingDashboard({ accountId: ':accountId' })} pathParams={{ accountId: 'dummy' }}>
        <LandingDashboardDeploymentsWidget />
      </TestWrapper>
    )

    // no deployments should be visible
    expect(getByText('No Deployments')).toBeTruthy()
  })
})
