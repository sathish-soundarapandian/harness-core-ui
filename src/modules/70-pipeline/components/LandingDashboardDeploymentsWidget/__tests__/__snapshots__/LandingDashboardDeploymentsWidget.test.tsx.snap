// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LandingDashboardDeploymentsWidget tests deployment badges 1`] = `
<div
  class="bp3-popover-content"
>
  <table
    class="bp3-html-table bp3-small popoverTable"
  >
    <thead>
      <tr>
        <th>
          PROJECTSTEXT
        </th>
        <th>
          DEPLOYMENTSTEXT
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <div
            class="Layout--vertical StyledProps--main"
          >
            <p
              class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--color StyledProps--color-grey1000"
              style="max-width: 200px; --text-line-clamp: 1;"
            >
              test1
            </p>
          </div>
        </td>
        <td>
          <p
            class="StyledProps--font StyledProps--main executionCount StyledProps--color StyledProps--color-primary7"
          >
            1
          </p>
        </td>
      </tr>
      <tr>
        <td>
          <div
            class="Layout--vertical StyledProps--main"
          >
            <p
              class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--color StyledProps--color-grey1000"
              style="max-width: 200px; --text-line-clamp: 1;"
            >
              test2
            </p>
          </div>
        </td>
        <td>
          <p
            class="StyledProps--font StyledProps--main executionCount StyledProps--color StyledProps--color-primary7"
          >
            2
          </p>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
